lab4.md

```diff
@@ -158,6 +158,65 @@ In Python code you'd type in `os.getenv("Http_X_Output_Mode")`.
 
 You can see that all other HTTP context is also provided such as `Content-Length` when the `Http_Method` is a `POST`, the `User_Agent`, Cookies and anything else you'd expect to see from a HTTP request.
 
+### Making use of logging
+
+The way OpenFaaS watchdog works is that it passses in the HTTP request and reads a HTTP response via the standard I/O streams `stdin` and `stdout` accordingly.
+This means that your process does not need to know anything about the web or HTTP.
+
+An interesting case is when it finishes with a non-zero exit code and `stderr` is not empty.
+By default function `stdout/stderr` is combined and `stderr` is not printed to the logs.
+
+Using the `combine_output` flag you can write `stderr` to the container logs and use only `stdout` for function response.
+
+Lets make a test with the `hello-openfaas` function from Lab 3.
+
+Change the `handler.py` code to
+
+```python
+import sys
+
+def handle(req):
+
+    sys.stderr.write("This should be an error message.\n")
+    print("Hello World!")
```

Comment:
Let's try to remove all references to print and do a final `return` instead. Anything returned is printed by the index.py (which is hidden)

What you want to show is that the response when it has logging in the middle of it - is no longer valid JSON:

```
return JSON.dumps({"hello": "world })
```

This is invalid JSON:

```
This should be an error message
{"hello": "world"}
```